src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
I've dug out the original discussions; PR #136 refactored `SourceLocator`s so that they return `Reflection|null` instead of `LocatedSource|null` (i.e. you're suggesting reversing that change). We had issues with the original implementation, #123 and #43 were solved by this change. The reason we changed it is the alternative solution is to load the AST of the source in the `SourceLocator` to check to see if the identifier is in there anyway, before returning the `LocatedSource` (with the assertion that the source has been found)... it makes no sense to parse the AST twice, so we moved the entire functionality into the `SourceLocator` directly. We toyed with the idea of having `PotentiallyLocatedSource` and so on, but those ideas still resulted in flakey, unpredictable behaviour. I can't honestly remember the exact reasons and issues we encountered, but we kept encountering problems, and all those problems disappeared when we moved the "AST to Reflection" conversion into the `SourceLocator`s themselves.

Ignoring the proposals made here for a moment, let's focus on the existing implementation of the `TypeFinders`. The context (i.e. "what namespace are we in, and what are the imports") is an absolute requirement to reliably determine types. At the moment, this functionality is tightly coupled to the instantiation and use of `phpDocumentor\Reflection\Types\ContextFactory`. This is made possible easily in the `Find(Parameter|Property|Return)Type` trio because we store the `LocatedSource` within the `Reflection` class itself. It's not as easy in `FindTypeFromAst` as we're not attaching to anything; but it's a reasonable ask to just inject a `LocatedSource` instance, which results in consistent behaviour across the 4 finders. In reality, these four classes are nothing more than a thin encapsulation that results in using the functionality from `phpDocumentor\Reflection\TypeResolver`.

Okay; lets look back at the initial cause for this PR: **performance**. We already mused a little about performance in #131. What's interesting is that by opening this particular PR, we can see the original implementation for the type finders was not ideal. They're tightly coupled, and don't allow any change in behaviour. The upside of the type finders is that their usage is simple, and that's led to them (or rather, `FindTypeFromAst` being used independently from the reflections itself; purely because it's a neat little encapsulation. In turn, many parts of Better Reflection itself are very tightly coupled to these classes. It is indeed a red light; in order to make those classes "flexible" (so that we can inject a `ContextFactory`), as you say, it involves passing it down the whole stack. I can certainly understand the reasoning for wanting to therefore have the `Reflector` contain the `ContextFactory`, but as Marco already pointed out, this isn't an ideal solution either.

I wonder if time would be better spent taking a step back and re-applying this in a different way. Rather than changing anything external to the type finders, we could I think, in theory, replace the functionality directly in the type finders (which are already tight couplings, so it's not exactly going to make it worse). You could still create a "native" `ContextFactory`, but it would 1) still remain coupled to `phpDocumentor` components and 2) add the caching behaviour in by default for *all* reflections. In essence, this would result in 1) no improvement in coupling (not good) 2) no BC breaks or large refactoring (preferable) 3) potential improvement of performance (good).

Not to say we don't want to go ahead and refactor properly, but if we do then it must be done properly, in which case the `ContextFactory` (or `Context` perhaps?) would indeed need to be injected all the way down the stack; however, this would be a big BC break (like I said before, I'm not averse to bumping a major version to do so), but we just need to be careful to ensure a reasonable upgrade path (and provide documentation). Maybe there is a better alternative solution we've not thought of as well! In any eventuality, I don't think reversing the changes made in PR #136 is a good idea (we'll end up in more pain; unless we can find a better solution there - but it's not a straightforward change I expect). I guess the TL;DR is "all or nothing" with regards to fixing the endemic issues of the type finders here ;)